Add plotters.

fchirica opened the pull request:
Add option `chia plotters` to CLI.

fchirica committed:
Initial commit add plotters.

fchirica committed:
Lint.

altendky reviewed (changes requested):
There are a few copy/paste or typo issues to fix.  The rest are comments for consideration or discussion if you are interested.

fchirica committed:
Add progress for bladebit.

fchirica commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,296 @@
+
+
+def call_plotters(root_path, args):
+    plotters = argparse.ArgumentParser(description="Available plotters.")
```
I wasn't aware we use Click when I wrote it. Probably we should refactor it with Click at some point.

paninaro replied:
We can revisit this at another time. Resolving.

fchirica commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,296 @@
+            )
+        if option is Options.FARMERKEY:
+            parser.add_argument(
+                "farmerkey",
```
As far as I remember, chiapos needs both id and memo, and max/bladebit need farmerkey. Would it be better to make them optional, and let the plotters throw at runtime if those are not provided?

paninaro replied:
In @altendky's comment with the output of `chia potters chiapos --help`, the usage: line is incorrect. I was noticing this as well while trying out the new CLI options.

altendky replied:
In chiapos, `--plotid` and `--memo` are both optional options and for debugging only.  I think the only positional argument for any of these plotters is the output directory for BladeBit.  Everything else is flagged options.  Their dependence on keys is the same except for chiapos being willing to pull the farmer public key and the pool public key from ...  the config or wherever it gets it.

If I remember correctly, in plotman I did add checks to make sure we had the farmer public key if using madMAx or BladeBit.  Likewise for requiring either the pool public key or the pool contract address for them, and for all that only one of those two is provided.  But, it is a bit smelly with the redundant checks.  There are trade offs both ways.

```console
$ chia plots create --help
Usage: chia plots create [OPTIONS]

Options:
  -k, --size INTEGER              Plot size  [default: 32]
  --override-k                    Force size smaller than 32  [default: False]
  -n, --num INTEGER               Number of plots or challenges  [default: 1]
  -b, --buffer INTEGER            Megabytes for sort/plot buffer  [default:
                                  3389]

  -r, --num_threads INTEGER       Number of threads to use  [default: 2]
  -u, --buckets INTEGER           Number of buckets  [default: 128]
  -a, --alt_fingerprint INTEGER   Enter the alternative fingerprint of the key
                                  you want to use

  -c, --pool_contract_address TEXT
                                  Address of where the pool reward will be
                                  sent to. Only used if alt_fingerprint and
                                  pool public key are None

  -f, --farmer_public_key TEXT    Hex farmer public key
  -p, --pool_public_key TEXT      Hex public key of pool
  -t, --tmp_dir PATH              Temporary directory for plotting files
                                  [default: .]

  -2, --tmp2_dir PATH             Second temporary directory for plotting
                                  files

  -d, --final_dir PATH            Final directory for plots (relative or
                                  absolute)  [default: .]

  -i, --plotid TEXT               PlotID in hex for reproducing plots
                                  (debugging only)

  -m, --memo TEXT                 Memo in hex for reproducing plots (debugging
                                  only)

  -e, --nobitfield                Disable bitfield
  -x, --exclude_final_dir         Skips adding [final dir] to harvester for
                                  farming

  -h, --help                      Show this message and exit.
```

paninaro replied:
+1 to @altendky's suggestion of making plotid and memo optional. Currently I'm required to enter memo and id params and it's unclear how to not specify them. Empty "" values?

paninaro commented on chia/cmds/plotters.py:
```diff
@@ -0,0 +1,9 @@
+from chia.plotters.plotters import call_plotters
+
+
+@click.command("plotters", context_settings={"ignore_unknown_options": True}, add_help_option=False)
```
Can we use `@click.group("plotters", short_help="Advanced plotting options")` (or some other helpful text)? Running `chia -h` shows the plotters command, but doesn't have any help text next to is

paninaro replied:
`Usage: chia [OPTIONS] COMMAND [ARGS]...

  Manage chia blockchain infrastructure (1.2.6.dev40)

Options:
  --root-path PATH  Config file root  [default: /Users/jeffc/.chia/mainnet]
  -h, --help        Show this message and exit.

Commands:
  configure   Modify configuration
  farm        Manage your farm
  init        Create or migrate the configuration
  keys        Manage your keys
  netspace    Estimate total farmed space on the network
  plotnft     Manage your plot NFTs
  plots       Manage your plots
  plotters
  run_daemon  Runs chia daemon
  show        Show node information
  start       Start service groups
  stop        Stop services
  version     Show chia version
  wallet      Manage your wallet`

paninaro commented on chia/plotters/bladebit.py:
```diff
@@ -0,0 +1,166 @@
+    if not os.path.exists(root_path / "bladebit/.bin/release/bladebit"):
+        print("Installing bladebit plotter.")
+        try:
+            install_bladebit(root_path)
```
I think it would be nice to have a separate 'install' command for bladebit/madmax. We could leave this logic as-is and attempt to install when attempting to start plotting, but it would be nice to allow bootstrapping with a dedicated install command. Something like:

`chia plotters install bladebit`

fchirica committed:
Address some review comments.

fchirica committed:
Oops...

fchirica committed:
Add install option.

paninaro commented on chia/plotters/bladebit.py:
```diff
@@ -0,0 +1,166 @@
+        try:
+            install_bladebit(root_path)
+        except Exception as e:
+            print(f"Exception while installing madmax plotter: {e}")
```
Copy & paste error:
`print(f"Exception while installing bladebit plotter: {e}")`

altendky replied:
Just to get it in the easily applied suggestion form.

```suggestion
            print(f"Exception while installing bladebit plotter: {e}")
```

paninaro commented on chia/plotters/install_plotter.py:
```diff
@@ -0,0 +1,31 @@
+            try:
+                install_bladebit(root_path)
+            except Exception as e:
+                print(f"Exception while installing madmax plotter: {e}")
```
maxmax -> bladebit

fchirica committed:
Update chia/plotters/bladebit.py

Co-authored-by: Jeff Cruikshank <paninaro@gmail.com>

fchirica committed:
Update chia/plotters/install_plotter.py

Co-authored-by: Jeff Cruikshank <paninaro@gmail.com>

fchirica committed:
Lint.

fchirica committed:
Change chiapos to fit the old standard.

fchirica committed:
Remove farmerkey as required.

fchirica committed:
Chia plotters chiapos works with no arguments.

paninaro commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,346 @@
+                default=None,
+                help="Enter the alternative fingerprint of the key you want to use",
+            )
+        if option is Options.CONNECT_TO_DAEMON:
```
TL;DR we don't need to expose this option. It was added as a hidden option just to support the GUI when a keyring passphrase is set. Users aren't likely to know when this option should be used.


The user experience when using the new passphrase-protected keyring is that CLI operations will prompt for the keyring passphrase to access keys. So running `chia plotters chiapos ...` without the `--connect_to_daemon` option will cause the keyring code to prompt for the user's passphrase. When plotting from the GUI, the user is asked to unlock their keyring at launch, and then the daemon maintains access to the keyring for the rest of the session. So when plotting from the GUI, we programmatically set the connect to daemon option '-D' so that the passphrase doesn't need to be re-entered.

paninaro committed:
Added get_plotters RPC to support the GUI (#8530)

* Added 'get_plotters' daemon RPC. Probes installed/available plotters on behalf of the GUI.

* Linter fix

* Minor type tweak

fchirica committed:
Tweaks.

fchirica committed:
Run with default arguments all plotters.

fchirica committed:
Fix bug.

paninaro commented on chia/plotters/bladebit.py:
```diff
@@ -0,0 +1,246 @@
+    for arg in args:
+        cmd += arg
+        cmd += " "
+    print(f"Running command: {cmd[:-1]}")
```
Why is the last arg left out?

fchirica replied:
it's a space character.

paninaro replied:
minor: could use `" ".join(args)` which would avoid the trailing space.

paninaro replied:
Good point @altendky re: escaping and params with spaces. Would subprocess.run() be a safer option?

altendky replied:
As noted in https://github.com/Chia-Network/chia-blockchain/pull/8497/files/4daee91fa3e296745cc09ee6bb481e013ee3dfd9#r713366777, we probably shouldn't be using shells.  And yeah, part of that is not joining the arguments into a single string but instead keeping them in a list.  For this print, it's just a matter of printing the output confusingly.  For the actual execution, yes, we should use the existing `args` list without having joined it.

Note though that `subprocess.run()` doesn't clarify between using a shell or not.  That's the `shell=` option to `.run()`.  Also, this is async and so needs to use an sync call, not the `subprocess` module directly.

paninaro replied:
@altendky I believe this has been addressed by @fchirica's refactoring. I tried plotting with some temp/final paths that included spaces in the name. Worked fine.

altendky replied:
Agreed.  I should be keeping on top of the changes and resolving my discussions.  Thanks for getting me back here.

paninaro commented on chia/plotters/bladebit.py:
```diff
@@ -0,0 +1,246 @@
+
+
+def plot_bladebit(args, chia_root_path, root_path):
+    if not os.path.exists(root_path / "bladebit/.bin/release/bladebit"):
```
```suggestion
    if not os.path.exists(get_bladebit_executable_path(root_path)):
```

fchirica committed:
Update chia/plotters/bladebit.py

Co-authored-by: Jeff Cruikshank <paninaro@gmail.com>

paninaro commented on chia/plotters/bladebit.py:
```diff
@@ -0,0 +1,246 @@
+        try:
+            subprocess.run(
+                [
+                    "sudo",
```
It occurs to me that we'll have trouble with this step if plotting from the GUI starts the installation step

altendky replied:
There are utilities to handle GUI prompts if appropriate.  I can't say I know the 'proper' option at this point.  But yes, handling installation via other means would also be good.

paninaro replied:
The issue is mostly resolved by having us package the prebuilt plotters for our installers, but there's still a question of how to handle this for users that clone the repo and build themselves. I'm leaning toward removing the option of installing the plotters on-demand when invoked from the GUI, but we could still build/install the plotters as part of install.sh and/or pull the prebuilt plotters. Basically anything that doesn't rely on sudo.

altendky replied:
I swear I don't mean to be nagging for installation from the GUI `:]`, if you are installing via a git clone then you are generally expected to be working from the CLI as well.  If you are using the package installers then certainly including prebuilt plotter binaries would be super nice.

But, it's not clear to me why the GUI itself is a major defining factor in this.  I think [`pkexec`](https://www.freedesktop.org/software/polkit/docs/latest/pkexec.1.html) is what I was thinking of for subprocessing with elevated privileges with a GUI password prompt.

paninaro replied:
@altendky I didn't find a pkexec or gksudo option for macOS. Is there a cross-plat option that you're aware of? Could use AppleScript (via osascript) to handle macOS I suppose.

My knee-jerk reaction against handling sudo-based install from the GUI is based on a few of things...

1) Source-based installation is not going to be bulletproof. We'll have all sorts of OS versions attempting to build the plotters, with varying library support (I assume hardware support isn't as critical though). I view the build process as having a high probability of failure, and I doubt we'll want to take on ownership of ensuring that the plotters can be built across supported OS versions (beyond our own ability to build them for packaging).

2) The UX for supporting install could be as simple as clicking a button and just waiting several minutes without any feedback, but I think that would be a poor experience. Marginally better would be to show a spinner to at least communicate start/finished status, but that still strikes me as a poor experience. To give the user some indication of how far along installation is, and really, to communicate errors, I think we would need to pipe the output to a log and show in the GUI. This is just a bit more work than I think we've signed on for at the moment. And even then, I'm not sure there's a way to provide the user with actionable advice when a build failure occurs.

3) I'll admit that I'm viewing this through a macOS-centric lens, but having having Python processes invoke a admin privileges prompt is something I find to be highly suspicious. As a user, I'd feel more comfortable with a native helper app (code signed) that performs the installation, but this would be a fair amount of work to build/package.

That said, this is just how I personally feel about privileged script-based installation, but I'm flexible and open to being persuaded :)

altendky replied:
Nope, I'm not pushing for taking the time to make the GUI install work and build out that experience.  I agree with your comments.  Comfort with the CLI is likely needed to address many potential issues with the builds and that does reduce the value of investing in putting this into the GUI.

I didn't look hard but along with `osascript` I saw reference to `cocoasudo`.

fchirica committed:
Change bladebit repo.

paninaro commented on chia/plotters/bladebit.py:
```diff
@@ -0,0 +1,246 @@
+        )
+    )
+    call_args = []
+    call_args.append(str(root_path) + "/bladebit/.bin/release/bladebit")
```
```suggestion
    call_args.append(str(get_bladebit_executable_path(root_path)))
```

fchirica committed:
Update chia/plotters/bladebit.py

Co-authored-by: Jeff Cruikshank <paninaro@gmail.com>

paninaro commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,359 @@
+            parser.add_argument(
+                "-w",
+                "--waitforcopy",
+                type=bool,
```
```suggestion
                action="store_true",
```

paninaro commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,359 @@
+                "--waitforcopy",
+                type=bool,
+                help="Wait for copy to start next plot",
+                default=True,
```
```suggestion
                action="store_true",
                help="Wait for copy to start next plot",
                default=False,
```

paninaro commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,359 @@
+            parser.add_argument(
+                "-w",
+                "--waitforcopy",
+                type=bool,
+                help="Wait for copy to start next plot",
+                default=True,
```
```suggestion
                action="store_true",
                help="Wait for copy to start next plot",
                default=False,

paninaro replied:
For the boolean options, the parser is currently expecting an argument. To keep the CLI similar to `chia plots create`, I'm thinking we can keep the boolean options standalone without requiring a 'true' or 'false' value (actually it's not clear what values are allowed, yes/no, y/n, true/false, 1/0, capitalized or not?)

In the case of `--waitforcopy` I think the default should be False (which would indicate that --waitforcopy wasn't specified in the command line). Personally I like having this option enabled, so I would always set it but perhaps I'm in the minority.

paninaro commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,359 @@
+        if option is Options.MADMAX_TMPTOGGLE:
+            parser.add_argument(
+                "-G",
+                "--tmptoggle",
+                help="Alternate tmpdir/tmpdir2 (default = false)",
```
```suggestion
                "--tmptoggle",
               action="store_true",
                help="Alternate tmpdir/tmpdir2 (default = false)",
```

paninaro replied:
"--tmptoggle",
                action="store_true",
                help="Alternate tmpdir/tmpdir2 (default = false)",

paninaro commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,359 @@
+            parser.add_argument(
+                "-e",
+                "--nobitfield",
+                type=bool,
```
```suggestion
                action="store_true",
```

paninaro commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,359 @@
+            parser.add_argument(
+                "-w",
+                "--warmstart",
+                type=bool,
+                help="Warm start (bool)",
```
```suggestion
                action="store_true",
                help="Warm start",
```

paninaro commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,359 @@
+            parser.add_argument(
+                "-m",
+                "--nonuma",
+                type=bool,
```
```suggestion
                action="store_true",
```

paninaro commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,359 @@
+            parser.add_argument(
+                "-v",
+                "--verbose",
+                type=bool,
```
```suggestion
                action="store_true",
```

paninaro commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,359 @@
+            )
+        if option is Options.OVERRIDE_K:
+            parser.add_argument(
+                "--override",
+                type=bool,
```
```suggestion
                "--override-k",
                action="store_true",
```

paninaro replied:
If we keep it as `--override-k` existing scripts/tools won't need to update.

paninaro commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,359 @@
+        if option is Options.EXCLUDE_FINAL_DIR:
+            parser.add_argument(
+                "-x",
+                "--exclude_final_dir",
+                help="Skips adding [final dir] to harvester for farming",
```
```suggestion
                "--exclude_final_dir",
                action="store_true",
                help="Skips adding [final dir] to harvester for farming",
```

fchirica committed:
Update chia/plotters/plotters.py

Co-authored-by: Jeff Cruikshank <paninaro@gmail.com>

fchirica committed:
Update chia/plotters/plotters.py

Co-authored-by: Jeff Cruikshank <paninaro@gmail.com>

fchirica committed:
Update chia/plotters/plotters.py

Co-authored-by: Jeff Cruikshank <paninaro@gmail.com>

fchirica committed:
Update chia/plotters/plotters.py

Co-authored-by: Jeff Cruikshank <paninaro@gmail.com>

fchirica committed:
Update chia/plotters/plotters.py

Co-authored-by: Jeff Cruikshank <paninaro@gmail.com>

fchirica committed:
Update chia/plotters/plotters.py

Co-authored-by: Jeff Cruikshank <paninaro@gmail.com>

fchirica committed:
Update chia/plotters/plotters.py

Co-authored-by: Jeff Cruikshank <paninaro@gmail.com>

fchirica committed:
Update chia/plotters/plotters.py

Co-authored-by: Jeff Cruikshank <paninaro@gmail.com>

paninaro committed:
Re-added the connect-to-daemon hidden option

altendky commented on chia/plotters/bladebit.py:
```diff
@@ -0,0 +1,246 @@
+                    "git",
+                ]
+            )
+        except Exception:
+            raise ValueError("Could not install dependencies.")
```
Report the error for consistency.

```suggestion
        except Exception as e:
            raise ValueError(f"Could not install dependencies. {e}")
```

altendky commented on chia/plotters/madmax.py:
```diff
@@ -0,0 +1,246 @@
+
+        try:
+            subprocess.run(["git", "--version"])
+        except FileNotFoundError as e:
+            raise ValueError(f"Git not installed. Aborting madmax install. {e}")
```
In general we seem to also catch 'all' errors.  For consistency...

```suggestion
        except FileNotFoundError as e:
            raise ValueError(f"Git not installed. Aborting madmax install. {e}")
        except Exception as e:
            raise ValueError(f"Error checking Git version. {e}")
```

altendky commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,372 @@
+                help="Stripe size.",
+                default=0,
+            )
+        if option is Options.TMP_DIR:
+            parser.add_argument(
+                "-t",
+                "--tmpdir",
+                type=str,
+                help="Temporary directory 1.",
+                default=str(root_path) + "/",
+            )
+        if option is Options.TMP_DIR2:
+            parser.add_argument(
+                "-2",
+                "--tmpdir2",
+                type=str,
+                help="Temporary directory 2.",
+                default=str(root_path) + "/",
+            )
```
These match madMAx right now instead of matching `chia plots create`.  Is this what we want?  I don't particularly like the `_` separators in our options, but consistency with ourselves does seem valuable.

paninaro replied:
Agreed that we should keep the options consistent with `chia plots create` to preserve functionality with existing user scripts/tools. So this would be changed to `--tmp2_dir` and we would also update `Options.FINAL_DIR` to `--final_dir`

altendky reviewed (dismissed):
Is `ValueError` something that is specifically caught and handled by our CLI framework?  It's not obvious to me that most of the exceptions ought to be `ValueError`s.

Also, subprocess failures are not being caught presently.  To do this without explicitly checking the return value of `subprocess.run()` you must pass `check=True` to get it to check itself.  (this really ought to be Python's default...  but oh well)

It might be worth just making a subprocess helper function to guarantee consistent exception handling and reporting as well as the `check=True` and anything else we come up with.  Presumably it would take the command and the 'nice' exception message like `run_the_process(command=["git", "clone", "something"], error_message="Unable to clone something repository.")`.

paninaro committed:
Use connect_to_daemon value for madmax and bladebit plot key resolution.

paninaro committed:
Updated --tmp_dir, --tmp_dir2, --final_dir options to match old options from chia plots create

paninaro committed:
Add CONNECT_TO_DAEMON as a valid option for madmax/bladebit

paninaro committed:
Thread multiplier should be an int

paninaro committed:
Passing params for madmax to start_plotting. Still needs cleanup/refactoring.

fchirica committed:
Update chia/plotters/bladebit.py

Co-authored-by: Kyle Altendorf <sda@fstab.net>

fchirica committed:
Update chia/plotters/madmax.py

Co-authored-by: Kyle Altendorf <sda@fstab.net>

fchirica committed:
Filename option -z.

fchirica commented on chia/plotters/plotters.py:
```diff
@@ -0,0 +1,372 @@
+            )
+        if option is Options.FILENAME:
+            parser.add_argument(
+                "--filename",
```
I added `-z` from lack of imagination, let me know if you have a better idea.

altendky replied:
Now that I think about this, what is this option for?

fchirica replied:
it's for having a custom filename of the final plot. (all of `c` for custom, `n` for name, `f` for filename are already taken..)

altendky replied:
Am I overlooking where it is used?  Or not implemented yet?

altendky replied:
I'm not sure about a full on custom name, but now that I think about it, adding extra tags for NFT plots or custom tags from users would be nice.  That's feature creep, not code quality though.  `:]`  Something for later.

fchirica replied:
this is an option for chiapos library, but it seems `chia plots create` indeed didn't leverage it at all. I removed it for now.

fchirica commented on chia/plotters/bladebit.py:
```diff
@@ -0,0 +1,246 @@
+
+
+async def run(command):
+    process = await asyncio.create_subprocess_shell(
```
I'm willing to use anything as long as we can redirect stdout and err to a`PIPE`. The example I found online that parses in real time stdout/err used subprocess shell, so I kept it.

altendky replied:
There's https://docs.python.org/3.9/library/asyncio-subprocess.html#asyncio.create_subprocess_exec.

fchirica committed:
Factor out calling the plotter.

fchirica committed:
First attempt refactor install scripts.

fchirica committed:
Switch to exec.

fchirica committed:
Attempt to fix mypy warning.

fchirica committed:
Remove filename.

paninaro committed:
Increase RLIMIT_NOFILE for madmax on non-Windows platforms

paninaro committed:
Add trailing path separator to madmax tmpdir/tmp2dir/finaldir arguments (required by madmax)

paninaro committed:
Fixes to support madmax plotting from the GUI.
Writing output from the plotters now includes a flush to ensure the plotter log (used by the GUI) is updated frequently.

paninaro committed:
Handle madmax's tmptoggle option internally when plotting with the GUI

paninaro committed:
Formatting and linter fix

mariano54 commented:
Should we add a test to ensure that these plotters are installing properly and running?

paninaro committed:
Fixed the -i option for bladebit

paninaro committed:
Construct BladeBit plotting options

paninaro committed:
Cleanup code for building plotter command line options

paninaro committed:
Added a post-processing step after each plotting job completes. Adds the final_dir plot directory as necessary.

cmmarslender committed:
Test adding mac madmax plotter to the installers

paninaro committed:
Fix plotter root path

paninaro committed:
Reverting prior checkin. Need to figure out how to handle CHIA_ROOT being overridden

paninaro committed:
BladeBit support for Windows

paninaro committed:
BladeBit's --memory-json option is used to check memory requirements

paninaro committed:
Madmax Windows support

paninaro committed:
Plotters directory is now under CHIA_ROOT

paninaro committed:
Madmax version detection

paninaro committed:
BladeBit will default to 0 threads. BladeBit will max-out available threads with this configuration.

paninaro committed:
LGTM fixes

altendky commented on chia/plotters/bladebit.py:
```diff
@@ -0,0 +1,199 @@
+        info["can_install"] = supported
+
+    if supported and meets_memory_requirement(plotters_root_path) is False:
+        info["bladebit_memory_warning"] = f"BladeBit requires at least {BLADEBIT_MIN_RAM_GIBS} GiB of RAM to operate"
```
BladeBit now outputs the minimum amount so we don't need to hardcode `BLADEBIT_MIN_RAM_GIBS`.

paninaro replied:
Done

altendky commented on chia/plotters/bladebit.py:
```diff
@@ -0,0 +1,199 @@
+        try:
+            proc = subprocess.run(
+                [os.fspath(get_bladebit_executable_path(plotters_root_path)), "--memory-json"],
+                capture_output=True,
```
Or maybe use the existing `run()` wrapper.

```suggestion
                capture_output=True,
                check=True,
```

paninaro replied:
Updated the remaining subprocess.run() calls to use run_command(). Required a few changes to run_command() but operates the same as before.

altendky commented on chia/plotters/plotters_util.py:
```diff
@@ -0,0 +1,48 @@
+        ]
+    )
+
+    await process.wait()
```
#### --- 1

I'm not clear after a quick look over the docs as to what happens if a process fails and returns a non-zero exit code.  Does our code presently become aware of that?  Do we notify the user?  The docs didn't even make it clear if we could pass `check=True` to `asyncio.create_subprocess_exec()`...  `:|`  Otherwise I guess we just manually check the [`process.returncode`](https://docs.python.org/3.9/library/asyncio-subprocess.html#asyncio.asyncio.subprocess.Process.returncode).


#### --- 2

But, the docs did yield something I wasn't looking for.

https://docs.python.org/3.9/library/asyncio-subprocess.html#asyncio.asyncio.subprocess.Process.wait
> Note This method can deadlock when using `stdout=PIPE` or `stderr=PIPE` and the child process generates so much output that it blocks waiting for the OS pipe buffer to accept more data. Use the [`communicate()`](https://docs.python.org/3.9/library/asyncio-subprocess.html#asyncio.asyncio.subprocess.Process.communicate) method when using pipes to avoid this condition.

I didn't check in detail but I think the stream reading code that runs before this would do so until the streams are closed thus leaving no remaining opportunity for this deadlock to occur.  But, perhaps it would be just as well to use `.communicate()` instead.

paninaro replied:
Regarding your first point, I tested a failed command that uses run_command, and indeed the exception is caught and our `exc_description` is printed, along with the command invocation and exit code.

I haven't observed any problems with the stream handling code while plotting with either plotter.

paninaro committed:
Module definition for chia.plotters to resolve mypy issues with chiapos (package vs chiapos.py)

paninaro committed:
Updated BladeBit build script to account for 1.2.0 changes.
Replaced remaining subprocess.run calls with run_command.
Use BladeBit's reported memory requirement instead of hardcoded value.

paninaro committed:
Show a disclaimer when using thirdparty plotter

cmmarslender committed:
Get latest madmax from the latest GH release

cmmarslender committed:
Fix bad var name

cmmarslender committed:
m1 madmax

cmmarslender committed:
Add linux/linux arm

cmmarslender committed:
pip install -e for arm installer, so that its consistent with the other platforms when looking for additional files

cmmarslender committed:
madmax + windows

cmmarslender committed:
Get madmax with Invoke-WebRequest

cmmarslender committed:
Use the correct windows slashes

cmmarslender committed:
add madmax to the list of windows binaries

cmmarslender committed:
Check if madmax exists on windows install and move it to site packages if it does

cmmarslender committed:
Make sure windows has .exe extension for madmax

cmmarslender committed:
Update azure to get latest version of madmax from GH releases

cmmarslender committed:
Bladebit for linux/linux arm

cmmarslender committed:
Fix error with binaries.extend

cmmarslender committed:
Bundle bladebit + windows

cmmarslender committed:
Fix download url for bladebit

cmmarslender committed:
Check for bladebit in windows script. move to correct directory if it exists

paninaro committed:
Detect and use packaged plotters

paninaro committed:
Removed unnecessary import

paninaro committed:
Updating the branch to use chiaplotters_gui for installer verification

paninaro committed:
Removed a change that was intended for debugging only

emlowe reviewed (dismissed):
ready to go

harold-b reviewed (dismissed):
Other than the `--no-cpu-affinity` which can be added in a different release, and the text edit, LGTM

paninaro committed:
Remove disclaimer

paninaro committed:
Updated for new madmax plotter with k33, k34 support.
Updated chia-blockchain-gui submodule

paninaro committed:
Fixed typo

paninaro committed:
Package the chia_plot_k34 executable

paninaro committed:
Boink

paninaro committed:
Revert "Boink"

This reverts commit 8d13c071100e99af8ec05ccb26eb2e9b411a6939.

paninaro committed:
Merge branch 'main' into fc.plotters

paninaro committed:
Additional chia_plot_k34 spots that I missed

paninaro committed:
pyinstaller.spec fix for chia_plot_k34

paninaro committed:
Windows installer fix for chia_plot_k34.exe

justinengland reviewed (dismissed):
provisionally approve the ci changes

paninaro committed:
Restoring chia-blockchain-gui submodule to 047ce16 (as in main)

emlowe committed:
Update to chiapos 1.0.6
